### PR TITLE
SDK-866: Introduce a docker environment for the Web server. Introduce SdkRestApiBundle with http endpoint.

### PR DESCRIPTION
**Pull request by Kyrylo-Meretskyi**
Tickets:
- [SDK-866](https://spryker.atlassian.net/browse/SDK-866)
- [SDK-865](https://spryker.atlassian.net/browse/SDK-865)

### CODE REVIEW

File: src/Presentation/RestApi/Service/RestApiTaskExecutor.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
+<?php
+
+/**
+ * Copyright © 2019-present Spryker Systems GmbH. All rights reserved.
+ * Use of this software requires acceptance of the Evaluation License Agreement. See LICENSE file.
+ */
+
+namespace SprykerSdk\Sdk\Presentation\RestApi\Service;
```
Comment (pavelmaksimov25) [inline src/Presentation/RestApi/Service/RestApiTaskExecutor.php:8]: Please, avoid using the `Service` folder. Use behaviour-specific naming instead. `Executor\Task` in this case.

File: docker-compose.rest-api.yaml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+            - php-fpm
+        restart: always
+
+    php-fpm:
+        container_name: sdk-php-fpm
+        build:
+            context: .
+            dockerfile: infrastructure/web/php-fpm.Dockerfile
+        volumes:
+            - ./:/data
```
Comment (pavelmaksimov25) [inline docker-compose.rest-api.yaml:22]: Must be specific per environment. `./:/data` binding ok for development, but for the prod mode, it shouldn't be like that, because the prod mode should use prebuild image instead and bind only specific files.

Please rename this file to `docker-compose.rest-api.dev.yaml` and create the separate task

File: docs/conventions.md [hunk 3]
```diff
@@ -89,6 +89,16 @@ No conventions yet
 - Contract __MUST__ exist only in case existing logic provides for an extension by the user.
 - Contract is a public API and __MUST__ follow [Spryker plugin interfaces specification](https://spryker.atlassian.net/wiki/spaces/RFC/pages/1038092073/INTEGRATED+RFC+Plugin+interface+specification).
 
+## REST API
```
Comment (pavelmaksimov25) [inline docs/conventions.md:92]: Please, move conventions to a separate PR. We need to think about it more carefully

Comment (Kyrylo-Meretskyi) [reply docs/conventions.md:92]: https://github.com/spryker-sdk/sdk/pull/247

File: src/Presentation/RestApi/Service/RestApiTaskExecutor.php (hunk 1 above)
Comment (Kyrylo-Meretskyi) [reply src/Presentation/RestApi/Service/RestApiTaskExecutor.php]: Fixed

File: docker-compose.rest-api.yaml (hunk 2 above)
Comment (Kyrylo-Meretskyi) [reply docker-compose.rest-api.yaml]: Fixed

**Review by sergeyspryker — CHANGES_REQUESTED** [review]

File: docker-compose.rest-api.dev.yaml [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        ports:
+            - '8080:80'
+        volumes:
+            - ./public:/data/public
+            - ./infrastructure/web/nginx/default.conf:/etc/nginx/conf.d/default.conf
+        depends_on:
+            - php-fpm
+        restart: always
+
+    php-fpm:
```
Comment (vol4onok) [inline docker-compose.rest-api.dev.yaml:16]: should have the correct permission. like for console build

File: src/Presentation/RestApi/Controller/v1/HelloController.php [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+use Symfony\Component\HttpFoundation\Response;
+
+class HelloController extends AbstractController
+{
+    /**
+     * @param \SprykerSdk\Sdk\Presentation\RestApi\Executor\Task\RestApiTaskExecutor $apiTaskExecutor
+     *
+     * @return \Symfony\Component\HttpFoundation\Response
+     */
+    public function helloWorld(RestApiTaskExecutor $apiTaskExecutor): Response
```
[the remark this answers is not in the document]
Comment (Kyrylo-Meretskyi) [reply src/Presentation/RestApi/Controller/v1/HelloController.php:21]: It's only for example that API can work with task execution. This controller will be deleted in future

File: infrastructure/web/php-fpm.dev.Dockerfile [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    && rm -rf /tmp/pear; apk del autoconf g++ make;
+
+COPY infrastructure/debug/php/69-xdebug.ini /usr/local/etc/php/conf.d/69-xdebug.ini
+
+RUN cp /usr/local/etc/php/php.ini-development /usr/local/etc/php/php.ini
+
+ARG USER_UID=1000
+ARG GROUP_UID=1000
+
+RUN deluser www-data && addgroup -g ${GROUP_UID} -S www-data && adduser -u ${USER_UID} -D -S -G www-data www-data
```
[the remark this answers is not in the document]
Comment (sergeyspryker) [reply infrastructure/web/php-fpm.dev.Dockerfile:19]: Alpine has no usermod

**Review by pavelmaksimov25 — COMMENTED** [review]
Please update the documentation with a rest-api approach usage info.

**Comment by vol4onok** [comment]
Needs to create Docker image build CI for API build.

File: src/Presentation/RestApi/Controller/v1/HelloController.php [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+     */
+    public function __construct(RestApiTaskExecutor $apiTaskExecutor)
+    {
+        $this->apiTaskExecutor = $apiTaskExecutor;
+    }
+
+    /**
+     * @return \Symfony\Component\HttpFoundation\Response
+     */
+    public function helloWorld(): Response
```
Comment (vol4onok) [inline src/Presentation/RestApi/Controller/v1/HelloController.php:32]: __invoke

File: src/Presentation/RestApi/Controller/v1/IndexController.php [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Component\HttpFoundation\Response;
+
+class IndexController
+{
+    /**
+     * @return \Symfony\Component\HttpFoundation\Response
+     */
+    public function index(): Response
```
Comment (vol4onok) [inline src/Presentation/RestApi/Controller/v1/IndexController.php:18]: __invoke

**Review by vol4onok — DISMISSED** [review]

File: docs/rest_api.md [hunk 9]
```diff
@@ -0,0 +1,33 @@
+# REST SDK API
+
```
Comment (pavelmaksimov25) [inline docs/rest_api.md:2]: We need to add a description here. 
So far, we can deliver it with the following text:
```
SDK Rest API enables an easy way to communicate with it via HTTP.
So far, it is an experimental feature and is available only for local development.
```

File: docs/conventions.md [hunk 10]
```diff
@@ -66,10 +66,14 @@ A placeholder must meet the following conditions:
 ## Console command
 A console command must meet the following conditions:
 
-- It shouldn't have any business logic.
-- Only basic input validation and output formatting should be present in the console command.
-- `protected static $defaultName` should not be used for performance reasons and future deprecation in Symfony version 6.1.
-- `protected const NAME` should be provided and passed to the parent constructor as a parameter instead of `protected static $defaultName`.
+- Console command __SHOULD__ have no business logic.
```
Comment (pavelmaksimov25) [inline docs/conventions.md:69]: I'm not sure we should overwrite console command conventions, as it's already edited by a technical writer.

File: docker-compose.rest-api.dev.yaml [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+        build:
+            context: .
+            dockerfile: infrastructure/web/nginx.Dockerfile
+        ports:
+            - ${SDK_REST_API_PORT:-80}:80
+        volumes:
+            - ./public:/data/public
+        depends_on:
+            - php-fpm
+        restart: always
```
Comment (pavelmaksimov25) [inline docker-compose.rest-api.dev.yaml:15]: Please replace it with restart: on-failure to prevent restarting loop.
The same for the prod env file.

File: docker-compose.rest-api.prod.yaml [hunk 12]
```diff
@@ -0,0 +1,21 @@
+version: '3.0'
```
Comment (pavelmaksimov25) [inline docker-compose.rest-api.prod.yaml:1]: so far we can remove prod file.

File: src/Presentation/RestApi/Controller/v1/HelloController.php (hunk 7 above)
Comment (pavelmaksimov25) [reply src/Presentation/RestApi/Controller/v1/HelloController.php:32]: agree. As we already agreed to go with single `__invoke()` method, let's stick to the agreement. @sergeyspryker, please change this controller and `IndexController`

File: src/Presentation/RestApi/Controller/v1/IndexController.php (hunk 8 above)
Comment (pavelmaksimov25) [reply src/Presentation/RestApi/Controller/v1/IndexController.php:18]: Also, it should return current version of SDK

File: src/Presentation/RestApi/Resources/config/routing.yaml [hunk 13]
```diff
@@ -0,0 +1,9 @@
+api_v1_index_index:
```
Comment (pavelmaksimov25) [inline src/Presentation/RestApi/Resources/config/routing.yaml:1]: incorrect route name. Must be `index`. Also, `api_v1` should be moved above as a route prefix. More info -> https://symfony.com/doc/current/routing.html#route-groups-and-prefixes

the same for the route below.

File: src/Presentation/RestApi/Controller/v1/HelloController.php [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+    {
+        $content = $this->apiTaskExecutor->execute([
+            'command' => 'hello:world',
+            '--world' => 'World',
+            '--somebody' => 'World',
+        ]);
+
+        return new JsonResponse(['result' => $content]);
```
Comment (pavelmaksimov25) [inline src/Presentation/RestApi/Controller/v1/HelloController.php:40]: Please, add at least some output structure. Currently, it looks weird, just a single line with `\n` as a string. Let's do a proper json.

**Review by pavelmaksimov25 — CHANGES_REQUESTED** [review]

File: docker-compose.rest-api.dev.yaml (hunk 11 above)
Comment (sergeyspryker) [reply docker-compose.rest-api.dev.yaml:15]: It doesn't prevent loop. But if the docker be restarted the services will not be upped.

**Review by pavelmaksimov25 — DISMISSED** [review]

**Review by pavelmaksimov25 — DISMISSED** [review]

**Review by sergeyspryker — DISMISSED** [review]

**Review by pavelmaksimov25 — DISMISSED** [review]

**Review by sergeyspryker — APPROVED** [review]
